### ISSUE TITLE
Failed to train Faster-rcnn on VOC2007 dataset

### ISSUE THREAD

**Issue by BIGJUN777** [pytorch/vision#1097]
Hello, 

These days, I try to train Faster-rcnn on VOC2007 dataset by transferring learning. The following code is what I used to construct the model: 
![Screenshot from 2019-07-06 17-10-03](https://user-images.githubusercontent.com/35102882/60754243-eb5a5f00-a010-11e9-8892-f32dab216f14.png)

And the main training code: 
![Screenshot from 2019-07-06 17-08-55](https://user-images.githubusercontent.com/35102882/60754236-c49c2880-a010-11e9-95ef-03a4b9fb64b3.png)

The model can be run successfully but not well. After several training epochs, I got the following 
graph of  loss
![Screenshot from 2019-07-06 16-39-16](https://user-images.githubusercontent.com/35102882/60754069-83a31480-a00e-11e9-88fe-b105e34f36b3.png)

When testing, however, the result was really bad: most of the labels of detection are wrong and the scores of interests are low. That says, it detected nothing:
![Screenshot from 2019-07-06 16-59-33](https://user-images.githubusercontent.com/35102882/60754148-776b8700-a00f-11e9-8d2a-639d89072f79.png)

I have tried to debug. Everything( the loaded images and label ) seems ok.  I think it is ok to just set the `num_classes` as 20, not including the `background`.  Is there something I missed? Does anyone have any idea on this?

Thanks.

**Comment by fmassa** [comment]
Hi,

You need to specify `num_classes + 1`, as the background class is not added automatically.

You should also check the learning rate and other hyperparameters, making sure that everything is setup properly.

Here is an example Pascal VOC class that I think worked fine
```python
import torch
import torchvision

import transforms as T


class PrepareInstance(object):
    CLASSES = (
        "__background__ ",
        "aeroplane",
        "bicycle",
        "bird",
        "boat",
        "bottle",
        "bus",
        "car",
        "cat",
        "chair",
        "cow",
        "diningtable",
        "dog",
        "horse",
        "motorbike",
        "person",
        "pottedplant",
        "sheep",
        "sofa",
        "train",
        "tvmonitor",
    )
    def __call__(self, image, target):
        anno = target['annotation']
        h, w = anno['size']['height'], anno['size']['width']
        boxes = []
        classes = []
        area = []
        iscrowd = []
        objects = anno['object']
        if not isinstance(objects, list):
            objects = [objects]
        for obj in objects:
            bbox = obj['bndbox']
            bbox = [int(bbox[n]) - 1 for n in ['xmin', 'ymin', 'xmax', 'ymax']]
            boxes.append(bbox)
            classes.append(self.CLASSES.index(obj['name']))
            iscrowd.append(int(obj['difficult']))
            area.append((bbox[2] - bbox[0]) * (bbox[3] - bbox[1]))

        boxes = torch.as_tensor(boxes, dtype=torch.float32)
        classes = torch.as_tensor(classes)
        area = torch.as_tensor(area)
        iscrowd = torch.as_tensor(iscrowd)

        image_id = anno['filename'][5:-4]
        image_id = torch.as_tensor([int(image_id)])

        target = {}
        target["boxes"] = boxes
        target["labels"] = classes
        target["image_id"] = image_id

        # for conversion to coco api
        target["area"] = area
        target["iscrowd"] = iscrowd

        return image, target


def get_voc(root, image_set, transforms):
    t = [PrepareInstance()]

    if transforms is not None:
        t.append(transforms)
    transforms = T.Compose(t)

    dataset = torchvision.datasets.VOCDetection(root, '2012', image_set, transforms=transforms)

    return dataset
```

I'm closing this issue as there is not much more information available so that I can give more advices.

**Comment by BIGJUN777** [comment]
Thanks. I will have a try.

**Comment by hktxt** [comment]
Hi, @BIGJUN777 ~
I have the same problem you post. 
Have you tried the suggestions and have a faithful result?

**Comment by BIGJUN777** [comment]
Hi, @hktxt 

I haven't tried the method above. Because I downloaded the dataset before and wrote a script to process the data from the original database. I just tried to modify some hyper-parameters to train. However, It turned out the same result. Maybe there is something wrong with my scripts. I still try to debug. If you get a good result with the above method. Please tell me.

Thanks!

**Comment by hktxt** [comment]
Okay~ @BIGJUN777 
The data processing script is not the point. more over, I use different backbones as posted here: https://pytorch.org/tutorials/intermediate/torchvision_tutorial.html#modifying-the-model-to-add-a-different-backbone and here https://github.com/hktxt/Faster-RCNN/blob/master/model/faster_rcnn.py#L86

`backbone = torchvision.models.mobilenet_v2(pretrained=True).features`  

`backbone.out_channels = 1280`  

`anchor_generator = AnchorGenerator(sizes=((128, 256, 512),), aspect_ratios=((0.5, 1.0, 2.0),))  `

`roi_pooler = torchvision.ops.MultiScaleRoIAlign(featmap_names=[0], output_size=7, sampling_ratio=2)`  

`model = FasterRCNN(backbone, num_classes=20, rpn_anchor_generator=anchor_generator, box_roi_pool=roi_pooler)`

However, trained failed. The loss vibrates, no convergence. The prediction means nothing:
![Adrian](https://user-images.githubusercontent.com/16555750/60786180-16b38a00-a188-11e9-9489-dc3702689432.jpg)

keep in touch

**Comment by BIGJUN777** [comment]
Hi, @hktxt 

The result I got is the same as yours. When testing, sometime it would show the wrong detection with a high score as the image you posted. Most of the time, however, it detected nothing. But as you can see the image of losses I posted above, The loss I got has converged. That's a little bit strange. BTW, how many epochs did you train the model to get the image above? Because you just used the pretrained backbone to construct the model, it may need time to train.

Keep in touch. Thanks.

**Comment by BIGJUN777** [comment]
Hi, @fmassa 

Sorry to bother you again.

Yesterday, I tried your code posted above. But it gave me the same result. The following are the main codes:

**model constructor:**

```
def Faster_RCNN(num_classes):

    anchor_genreator = AnchorGenerator(sizes=(128, 256, 512), aspect_ratios=(0.5, 1.0, 2.0))
    roi_pooler = torchvision.ops.MultiScaleRoIAlign(featmap_names=[0],
                                                    output_size=7,
                                                    sampling_ratio=2)
    model = torchvision.models.detection.fasterrcnn_resnet50_fpn(pretrained=True, 
                                                                 min_size=600, 
                                                                 max_size=1000,
                                                                #  rpn_anchor_generator=anchor_genreator,
                                                                 rpn_pre_nms_top_n_train=12000, 
                                                                 rpn_pre_nms_top_n_test=6000,
                                                                 rpn_post_nms_top_n_train=2000, 
                                                                 rpn_post_nms_top_n_test=300,
                                                                 rpn_nms_thresh=0.7,
                                                                 rpn_fg_iou_thresh=0.7, 
                                                                 rpn_bg_iou_thresh=0.3,
                                                                 rpn_batch_size_per_image=256, 
                                                                 rpn_positive_fraction=0.5,
                                                                 box_batch_size_per_image=128, 
                                                                 box_positive_fraction=0.25,
                                                                 box_score_thresh=0.1, 
                                                                 box_nms_thresh=0.3, 
                                                                 box_detections_per_img=100)

    model.rpn.anchor_genreator = anchor_genreator
    model.roi_heads.roi_pooler = roi_pooler
    in_features = model.roi_heads.box_predictor.cls_score.in_features
    model.roi_heads.box_predictor = FastRCNNPredictor(in_features, num_classes) 

    return model
```

**The dataset class is the same as you suggested. I just added some transformer into the *get_voc* function and used the 2007 version dataset(I have downloaded) instead of the 2012 version:**

```
def get_voc(root, image_set, transforms=None):
    t = [PrepareInstance()]

    if transforms is not None:
        t.append(transforms)
    t.append(T.ToTensor())
    t.append(T.RandomHorizontalFlip(0.5))
    transforms = T.Compose(t)

    dataset = torchvision.datasets.VOCDetection(root, '2007', image_set, transforms=transforms, download=False)

    return dataset
```

**Main training code:**

```
    dataset = get_voc('database', 'trainval', )
    dataloader = torch.utils.data.DataLoader(dataset, batch_size=args.batch_size, shuffle=True, collate_fn=collate_fn)
    print('data prepared, train data: %d' % len(dataset))
    model = Faster_RCNN(num_classes=21)
    device = torch.device('cuda:0' if torch.cuda.is_available() and args.gpu else 'cpu')
    model.to(device)
    print('model construct completed. Training on %s' % device)
    params = [p for p in model.parameters() if p.requires_grad]
    
    # optimizer = optim.Adam(model.parameters(), lr=0.005)
    optimizer = optim.SGD(params, lr=0.005, momentum=0.9, weight_decay=0.0005)
    lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=30, gamma=0.1)
    # setup log data writer
    if not os.path.exists('log'):
        os.makedirs('log')
    writer = SummaryWriter(log_dir='log')

    for epoch in range(300):
        loss_epoch = {}
        loss_name = ['loss_classifier', 'loss_box_reg', 'loss_objectness', 'loss_rpn_box_reg']
        for ii, (images, targets) in tqdm(enumerate(dataloader)):   
            model.train()
            model.zero_grad()
            images = list(image.to(device) for image in images)
            targets = [{k: v.to(device) for k, v in t.items()} for t in targets]
            # training
            loss_dict = model(images, targets)
            losses = sum(loss for loss in loss_dict.values())
            losses.backward()
            optimizer.step()
```

**And I got the following result. From the image, we can know the loss have converged after several epochs. When testing, however, it shows the wrong detection even nothing most of the time. Do you have any idea on this problem? I am really confused about this these days. Thinks.**

![Screenshot from 2019-07-09 10-04-19](https://user-images.githubusercontent.com/35102882/60853973-f55aa880-a230-11e9-9672-d7da1256b3c4.png)

![Screenshot from 2019-07-09 10-16-39](https://user-images.githubusercontent.com/35102882/60854758-f214ec00-a233-11e9-8ac6-ac4c963172af.png)

**Comment by hktxt** [comment]
@BIGJUN777 
I think it trainable using pretrained fasterrcnn_resnet50_fpn on coco. I use this pretrained model and trained on voc dataset for only 3 epochs with lr=1e-4. And I got a map 0.75. 
`100%|??????????| 4952/4952 [06:00<00:00, 13.73it/s]
0.7480373006085486`
but it gives too many predictions:
![000000000081](https://user-images.githubusercontent.com/16555750/60855037-d3fbbb80-a234-11e9-87db-f7f48b7d38de.jpg)
So I applied nms: https://github.com/pytorch/vision/blob/master/torchvision/ops/boxes.py#L5
![000000000081](https://user-images.githubusercontent.com/16555750/60855117-1b824780-a235-11e9-9f13-ca83e104f055.jpg)
works well.

I post my code in brief.

```
model = torchvision.models.detection.fasterrcnn_resnet50_fpn(pretrained=True)
in_features = model.roi_heads.box_predictor.cls_score.in_features
model.roi_heads.box_predictor = FastRCNNPredictor(in_features, num_classes=21)

train_set = get_voc(image_set='trainval')
test_set = get_voc(image_set='test')
train_loader = torch.utils.data.DataLoader(train_set, batch_size=1, num_workers=15)
test_loader = torch.utils.data.DataLoader(test_set, batch_size=1, num_workers=15)

params = [p for p in model.parameters() if p.requires_grad]
optimizer = torch.optim.SGD(params, lr=0.0001, momentum=0.9, weight_decay=0.005)

for epoch in range(3):
     for step, (img, targets) in enumerate(train_loader):
        tensor_image = list(img.to(device))
        tensor_targets = [{k: v[0].to(device) for k, v in targets.items()}]

        loss_dict = model(tensor_image, tensor_targets)  # images size changes
        losses = sum(loss for loss in loss_dict.values())

        rpn_objectness = loss_dict['loss_objectness']
        rpn_box_reg = loss_dict['loss_rpn_box_reg']
        roi_classifier = loss_dict['loss_classifier']
        roi_box_reg = loss_dict['loss_box_reg']

        optimizer.zero_grad()
        losses.backward()
        optimizer.step()
```

**Comment by fmassa** [comment]
@BIGJUN777 you changed the Anchor Generator of the faster rcnn model. This means that you also need to change the RPN Classifier, because they have been trained with FPN.

This is probably the issue.

Note that we only make available FPN models. If you want non-FPN models, you might need to adapt the heads so that they are more adapted (see the C4 resnet style for example)

**Comment by hktxt** [comment]
@fmassa  Hi~
I found that it's hard to train faster rcnn from scratch not using pretrained model.
I follow this [tutorial](https://pytorch.org/tutorials/intermediate/torchvision_tutorial.html#modifying-the-model-to-add-a-different-backbone), use different backbone to train, but always failed.
Does it need special treatment to train faster rcnn from scratch?
one more weird thing is that the [min_size/max_size](https://github.com/pytorch/vision/blob/master/torchvision/models/detection/faster_rcnn.py#L145) cann't set 600/1000, that will cause trian error.

**Comment by fmassa** [comment]
@hktxt training detection models from scratch takes a lot of time and data, that's probably why you are not able to get it to work.

> one more weird thing is that the min_size/max_size cann't set 600/1000, that will cause trian error.

What error do you get?

**Comment by BIGJUN777** [comment]
@hktxt  @fmassa Thank you for your help.

**Comment by hktxt** [comment]
@fmassa 
It's ok now.
@BIGJUN777 
have you reproduced the result which is post in the paper?
This is a generalized framework, which has some differences compared with original faster-rcnn.
I made some modifications to make sure it's exactly the same as the paper said.
I got mAP 61.1%, lower than reported 69.9%.
I think there are some differences that I neglected. 
If you are doing the same work, plz share it.
